core/arch/arm/plat-rockchip/platform_rk3588.c

```diff
@@ -160,3 +191,197 @@ TEE_Result hw_get_random_bytes(void *buf, size_t blen)
+		res = tee_otp_read_secure(&old_val, curr_idx, OTP_WORD);
+		if (res != TEE_SUCCESS)
+			return res;
+
+		/* Only program bits that are currently 0 (0->1) */
+		new_val = *pValue++ & ~old_val;
+		if (!new_val)
+			continue;
+
+		/* Clear any pending interrupts */
+		io_write32(base + OTP_S_INT_ST,
+			   io_read32(base + OTP_S_INT_ST));
+
+		/* Set write command */
+		io_write32(base + OTP_S_AUTO_CTRL, auto_ctrl_val);
+
+		/* Write the new bits into PROG_DATA register */
+		io_write32(base + OTP_S_PROG_DATA, new_val);
+
+		/* Enable the write */
+		io_write32(base + OTP_S_AUTO_EN, EN_ENABLE);
+
+		/* Poll for WR_DONE or verify/error bits */
+		res = IO_READ32_POLL_TIMEOUT(base + OTP_S_INT_ST,
+					     int_status,
+					     (int_status & OTP_S_WR_DONE_BIT) ||
+					     (int_status & OTP_S_VERIFY_BIT) ||
+					     (int_status & OTP_S_ERROR_BIT),
+					     OTP_POLL_PERIOD_US,
+					     OTP_POLL_TIMEOUT_US);
+
+		/* Check for VERIFY_FAIL, ERROR or timeout */
+		if (int_status & OTP_S_VERIFY_BIT) {
+			EMSG("OTP_S Verification fail");
+			return TEE_ERROR_GENERIC;
+		}
+		if (int_status & OTP_S_ERROR_BIT) {
+			EMSG("OTP_S Error");
+			return TEE_ERROR_GENERIC;
+		}
+		if (res) {
+			EMSG("OTP_S Timeout");
+			return TEE_ERROR_BUSY;
+		}
+
+		/* Clear INT status bits */
+		io_write32(base + OTP_S_INT_ST, int_status);
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
+{
+	TEE_Result res = 0;
+	uint32_t buffer[HW_UNIQUE_KEY_SIZE / sizeof(uint32_t)];
+	bool key_is_empty = true;
+
+	if (!hwkey)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(&huk_mutex);
+
+	/* Read 4 words (16 bytes) from OTP at HW_UNIQUE_KEY_INDEX */
+	memset(buffer, 0, sizeof(buffer));
+	res = tee_otp_read_secure(buffer, HW_UNIQUE_KEY_INDEX,
+				  HW_UNIQUE_KEY_SIZE / sizeof(uint32_t));
+	if (res != TEE_SUCCESS) {
+		mutex_unlock(&huk_mutex);
+		return res;
+	}
+
+	/* Check if the buffer is all zero => HUK not present */
+	for (size_t i = 0; i < ARRAY_SIZE(buffer); i++) {
+		if (buffer[i] != 0) {
+			key_is_empty = false;
+			break;
+		}
+	}
+
+	if (key_is_empty) {
+		/* HUK is empty => generate random key and program it */
+		uint8_t rand_bytes[HW_UNIQUE_KEY_SIZE];
+		uint32_t words[HW_UNIQUE_KEY_SIZE / sizeof(uint32_t)];
+
+		/* Generate random 128-bit key from TRNG */
+		res = hw_get_random_bytes(rand_bytes, sizeof(rand_bytes));
+		if (res != TEE_SUCCESS) {
+			mutex_unlock(&huk_mutex);
+			return res;
+		}
+
+		/* Convert random bytes -> 32-bit words */
+		memcpy(words, rand_bytes, sizeof(words));
+
+		/* Write the new HUK into OTP at HW_UNIQUE_KEY_INDEX */
+		res = tee_otp_write_secure(words, HW_UNIQUE_KEY_INDEX,
+					   ARRAY_SIZE(words));
+		if (res != TEE_SUCCESS) {
+			mutex_unlock(&huk_mutex);
+			return res;
+		}
+
+		/* Update buffer with the newly generated words */
+		memcpy(buffer, words, sizeof(buffer));
+	}
+
+	/* Copy HUK into hwkey->data */
+	memcpy(hwkey->data, buffer, HW_UNIQUE_KEY_SIZE);
```

Comment:
Since the HUK is a sensible data, I suggest to wipe `buffer content` (`memzero_explicit()`).
By the way, maybe using straight `hwkey->data` to get/generate the HUK is enough, no strict need to use a temporary buffer.